### ISSUE TITLE
[RFC] Better granularity in index for reuse of information

### ISSUE THREAD

**Issue by mosteo** [alire-project/alire#175]
Context: in the old index you could use a release as the base for a new one ([example](https://github.com/alire-project/old-index/blob/master/oldindex/alire-index-adayaml.ads)), either by replacing all or adding new dependencies/properties. This, to some extent, is doable in the new index because the `[general]` sections allows to specify properties/dependencies that apply to all releases in the crate.

I see a few shortcomings with the current status:
* If you specify some dependencies/properties in `[general]`, these will apply forever to all releases; so, in turn:
* You cannot in practice modify the [general] section once you have a few releases, since that would apply to past releases, or
* If you do want to modify the general section, you'll have to move its defaults to all existing releases, and start afresh.
* It seems unlikely to me that, in long-lived crates, with possibly several major point releases, `[general]` deps/props can be reused; in practice every release will have to list all its information, even if there is some extent of commonality.

TL;DR: some way of having more precise inheritance seems likely to be needed. I see a couple obvious possibilities, haven't given too much thinking to it, waiting for comments:
1. Be able to refer to an existing release, as in the past, to add new info incrementally.
1. Have some kind of "named sections" which are not releases to which a release can refer as a base; thus several releases may reuse a common section. Once breaking changes appear, a new common section is created as baseline, without the need to change anything anywhere.

And as how to express that, I can see:
1. Using explicit names and a way of referring to them (`inherits-from = "whatever"` or the like)
1. Have an extra nesting level in which several crates can be grouped within a common section. Trouble is that TOML is somewhat cumbersome for nested structures, but that cat is already out of the bag. If we where using YAML I would go for this one but with TOML I feel 1 may be clearer.

If some solution were adopted, I would suggest to reserve the `[general]` section for truly unique values (authors, maintainers) that once changed really apply to all crates, but not for deps/props.

**Comment by Fabien-Chouteau** [comment]
How is this done in other package managers?

**Comment by mosteo** [comment]
Rust: not done, apparently: https://github.com/rust-lang/crates.io-index/blob/master/ra/nd/rand

OPAM: not done, and besides they have one file per release: https://github.com/ocaml/opam-repository/tree/master/packages/mirage-random

I guess that if the ultimate goal is to only edit those files through the tooling it doesn't make sense to make them amenable to user edition (or even reading, as in the Rust case).
